src/mapgen.cpp

```diff
@@ -4479,59 +4468,87 @@ bool mapgen_function_json_base::setup_common( const JsonObject &jo )
                                total_size.y, parray.size() ) );
         }
-        for( int c = m_offset.y; c < expected_dim.y; c++ ) {
-            const std::string row = parray.get_string( c );
-            static std::vector<std::string_view> row_keys;
-            row_keys.clear();
-            row_keys.reserve( total_size.x );
-            utf8_display_split_into( row, row_keys );
-            if( row_keys.size() < static_cast<size_t>( expected_dim.x ) ) {
-                parray.throw_error(
-                    string_format( "  format: row %d must have at least %d columns, not %d",
-                                   c + 1, expected_dim.x, row_keys.size() ) );
-            }
-            if( row_keys.size() != static_cast<size_t>( total_size.x ) ) {
-                parray.throw_error(
-                    string_format( "  format: row %d must have %d columns, not %d; check mapgensize if applicable",
-                                   c + 1, total_size.x, row_keys.size() ) );
-            }
-            for( int i = m_offset.x; i < expected_dim.x; i++ ) {
-                const point p = point( i, c ) - m_offset;
-                const map_key key{ std::string( row_keys[i] ) };
-                const auto iter_ter = keys_with_terrain.find( key );
-                const auto fpi = format_placings.find( key );
-
-                const bool has_terrain = iter_ter != keys_with_terrain.end();
-                const bool has_placing = fpi != format_placings.end();
-
-                if( !has_terrain && !fallback_terrain_exists ) {
+    } else {
+        // Construct default "rows" of just spaces
+        const std::string spaces( expected_dim.x, ' ' );
+        std::string empty_row( 1, '"' );
+        empty_row += spaces;
+        empty_row += '"';
+        std::string empty_rows = "[";
+        for( int rown = 0; rown < expected_dim.y - 1; rown++ ) {
+            empty_rows += empty_row;
+            empty_rows += ',';
+        }
+        empty_rows += empty_row;
+        empty_rows += ']';
+        std::shared_ptr<parsed_flexbuffer> parsed_array_ptr = flexbuffer_cache::parse_buffer( empty_rows );
+        parray = JsonArray( parsed_array_ptr,
+                            flexbuffer_root_from_storage( parsed_array_ptr->get_storage() ), {} );
+    }
+
+    // just like mapf::basic_bind("stuff",blargle("foo", etc) ), only json input and faster when applying
+    mapgen_palette palette = mapgen_palette::load_temp( jo, "dda", context_ );
+    auto &keys_with_terrain = palette.keys_with_terrain;
+    mapgen_palette::placing_map &format_placings = palette.format_placings;
+
+    if( palette.keys_with_terrain.empty() && !fallback_terrain_exists ) {
+        return false;
+    }
+
+    parameters = palette.get_parameters();
+
+    for( int c = m_offset.y; c < expected_dim.y; c++ ) {
+        const std::string row = parray.get_string( c );
+        static std::vector<std::string_view> row_keys;
```

Comment:
It might end up simpler to do something like this.
```suggestion

    std::function<std::string(int)> get_row;
    std::function<void(std::string)> json_error;
    ...
    if ( jo.has_array( "rows" ) ) {
        parray = jo.get_array( "rows" );
        if( static_cast<int>( parray.size() ) < expected_dim.y ) {
            parray.throw_error( 
                string_format( "format: rows: must have %d rows, not %d; check mapgensize if applicable", total_size.y, parray.size() ) );
        get_row = [&]( int i ) {
          return parray.get_string( i );
        };
        json_error = [&]( std::string error ) {
          parray.string_error( error );
        }
    } else {
        // Construct default "rows" of just spaces
        const std::string spaces( expected_dim.x, ' ' );
        get_row = [=](int) {
          return spaces;
        };
        json_error = [](std::string error) {
          // won't happen.
        };
    }

    // just like mapf::basic_bind("stuff",blargle("foo", etc) ), only json input and faster when applying
    mapgen_palette palette = mapgen_palette::load_temp( jo, "dda", context_ );
    auto &keys_with_terrain = palette.keys_with_terrain;
    mapgen_palette::placing_map &format_placings = palette.format_placings;

    if( palette.keys_with_terrain.empty() && !fallback_terrain_exists ) {
        return false;
    }

    parameters = palette.get_parameters();

    for( int c = m_offset.y; c < expected_dim.y; c++ ) {
        const std::string row = parray.get_string( c );
        static std::vector<std::string_view> row_keys;
```